nmostesting/mocks/Registry.py

```diff
@@ -271,3 +417,192 @@ def heartbeat(version, node_id):
+                            and NMOSUtils.compare_resource_version(until, value['version']) > 0:
+                        if result_count < registry.paging_limit:
+                            base_data.append(value)
+                            result_count += 1
+                        else:
+                            new_until = value['version']
+                            break
+
+    except Exception:
+        pass
+
+    response = Response(json.dumps(base_data), mimetype='application/json')
+
+    # add pagination headers for v1.1 and up
+    if NMOSUtils.compare_api_version("v1.1", version) <= 0:
+        link = ""
+        if new_until != MAX_UNTIL:
+            link += "<http://localhost:5102/x-nmos/query/v1.3/" + resource_type + "s/?paging.since=" + new_until \
+                + "&paging.limit=" + str(registry.paging_limit) + ">; rel=\"next\""
+
+        if since != MIN_SINCE:
+            if link != "":
+                link += ","
+            link += "<http://localhost:5102/x-nmos/query/v1.3/" + resource_type + "s/?paging.since=0:0&paging.limit=" \
+                + str(registry.paging_limit) + ">; rel=\"first\""
+
+        if link != "":
+            response.headers["Link"] = link
+        response.headers["X-Paging-Limit"] = registry.paging_limit
+        response.headers["X-Paging-Since"] = since
+        response.headers["X-Paging-Until"] = new_until
+        response.headers["X-Ancestry-Generations"] = 1
+
+    return response
+
+
+@REGISTRY_API.route('/x-nmos/query/<version>/<resource>/<resource_id>', methods=['GET'], strict_slashes=False)
+def get_resource(version, resource, resource_id):
+    registry = REGISTRIES[flask.current_app.config["REGISTRY_INSTANCE"]]
+    if not registry.enabled:
+        abort(503)
+    authorized = registry.check_authorized(request.headers, request.path)
+    if authorized is not True:
+        abort(authorized)
+
+    registry.query_api_called = True
+
+    resource_type = resource.rstrip("s")
+    data = []
+    try:
+        data = registry.get_resources()[resource_type][resource_id]
+    except Exception:
+        abort(404)
+        pass
+
+    return Response(json.dumps(data), mimetype='application/json')
+
+
+@REGISTRY_API.route('/x-nmos/query/<version>/subscriptions', methods=["POST"])
+def post_subscription(version):
+
+    registry = REGISTRIES[flask.current_app.config["REGISTRY_INSTANCE"]]
+    if not registry.enabled:
+        abort(503)
+    authorized = registry.check_authorized(request.headers, request.path)
+    if authorized is not True:
+        abort(authorized)
+
+    subscription_request = request.json
+
+    subscription_response = {}
+    created = False
+
+    try:
+        subscription, created = registry.subscribe_to_query_api(version, subscription_request["resource_path"])
+
+        # Note: 'secure' not required in request, but is required in response
+        secure = subscription_request['secure'] if 'secure' in subscription_request else False
```

Comment:
All of the test suites can be run in secure comms mode, so I'm pretty sure the mock Registry supports HTTPS in the case. So this should reflect the config setting rather than `False`. Requests that don't match the cheng mode should be rejected.

Reply:
Hmmm, yes - there's a wider issue, which is that the Registry is running a WebSocket server separate to the Flask app which handles the subscriptions. This is fully capable of running in secure mode, and just needs to pick up the cert and key from the config (cf. the secure Flask app). However this needs to be tested with an appropriate secure controller, so in the interests of time I wonder whether this should be deferred as an issue, rather than solved in this PR?

Reply:
So if the testing tool is run in secure mode, what will happen for now?

If I've understood correctly, the WebSocket server is always run insecure?

I think it would be best to fix the issue _here_ now (i.e. default to `Config.ENABLE_HTTPS` rather than `False`), and to also ensure that if a secure subscription is created (whether by explicit request from the controller, or via this default), the relevant test case fails with an appropriate error - I think this ought to be by `Test.DISABLED` to indicate that this is a test suite problem not a fault with the NCuT.

Reply:
We should probably reject the subscription request if secure is `True `(assuming secure is only ever present and set to `True` when a wss WebSocket is expected) possibly with a 400 (yes?) or 500 (maybe?) code?

If a secure subscription has been attempted, and no insecure subscriptions have been subsequently attempted can we assume `Test.DISABLED` for any tests that require the query API (which, in fact, is all of them) - assuming the REST API isn't being used as fallback?

Reply:
Ok... yeah... given that the mock only supports insecure subscriptions for now, yes, rejecting requests for secure ones with `501 Not Implemented` is logically the right thing to do. (Still need the default fix though.)

As you say, this would allow a controller to fall back to insecure WebSocket. However, if it _didn't_, it's going to get a bunch of test failures... maybe, in an interactive session, a user would just bail out knowing they had no means of answering the following questions, hmm. Do we need to also bear in mind that falling back like this might be considered bad practice? Why use a secure HTTPS API to set up the subscription if you're going to exchange all subsequent data in the clear?

Reply:
Code modified according to above discussion, but wider discussion needed about permissiveness of subscription requests